geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/ConcatenatedFileWriter.scala

```diff
@@ -38,17 +38,31 @@ object ConcatenatedFileWriter {
       ipcOpts: IpcOption,
       sort: Option[(String, Boolean)],
       files: CloseableIterator[Array[Byte]]): CloseableIterator[Array[Byte]] = {
-    // ensure we return something
-    if (files.hasNext) { files } else {
-      val dictionaries = dictionaryFields.mapWithIndex { case (name, i) =>
-        name -> ArrowDictionary.create(sft.getTypeName, i, Array.empty[AnyRef])
+    // NB: This is not a WithClose situation.
+    //  If there is an empty/non-exceptional iterator, we wish to use it.
+    //  If there are any issues, we wish to close the iterator to free memory.
+    try {
+      if (files.hasNext) {
+        files
+      } else {
+        files.close() // Closing the input just to be sure.
+        generateEmptyResponse(sft, dictionaryFields, encoding, ipcOpts, sort)
       }
-      val os = new ByteArrayOutputStream()
-      WithClose(SimpleFeatureArrowFileWriter(os, sft, dictionaries.toMap, encoding, ipcOpts, sort)) { writer =>
-        writer.flush() // ensure header and dictionaries are written, and write an empty batch
-      }
-      // files is empty but this will pass it through to be closed
-      files ++ CloseableIterator.single(os.toByteArray)
+    } catch {
+      case t: Throwable =>
+        files.close()
+        throw t
+    }
+  }
+
+  private def generateEmptyResponse(sft: SimpleFeatureType, dictionaryFields: Seq[String], encoding: SimpleFeatureEncoding, ipcOpts: IpcOption, sort: Option[(String, Boolean)]) = {
+    val dictionaries = dictionaryFields.mapWithIndex { case (name, i) =>
```

Comment:
so technically the dictionaries should be closed - since we're creating them from an Array the close is a no-op. but maybe we should close them after closing the writer, just in case it changes in the future?

Reply:
Would a 
```dictionaries.foreach { case (_, dictionary) => dictionary.close() }``` fit the bill?

Reply:
probably want to do something like `CloseWithLogging(dictionaries.map(_._2))` to ensure they all get closed if one errors.